phylogeny/prototypeSelection.py

```diff
@@ -0,0 +1,106 @@
+"""
+Prototype selection
+
+Given a set of n elements (S) and their pairwise distances in terms of a
+dissimilarity matrix (DM) and a given number (k << n), find the sub-set (s)
+consisting of exactly k elements, such that s best represents the full set S.
+
+Here, we define "best represents" as maximizing the sum of distances between
+all points, i.e. sum {over a,b in S} DM[a,b]. This is our objective function.
+
+This problem is known to be NP-hard [1], thus we need to resort to heuristics.
+In the future, this module will implement several heuristics, whose quality can
+be measured by the objective function for each problem instance, since there is
+no global winner.
+For completeness, the exact but exponential algorithm is implemented, too.
+  "prototypeSelection_exhaustive"
+
+[1] Gamez, J. Esteban, François Modave, and Olga Kosheleva.
+    "Selecting the most representative sample is NP-hard:
+     Need for expert (fuzzy) knowledge."
+    Fuzzy Systems, 2008. FUZZ-IEEE 2008.
+"""
+
+from skbio.stats.distance import DistanceMatrix
+import itertools
+import numpy as np
+import scipy as sp
+
+
+def distanceSum(elements, dm: DistanceMatrix):
+    '''Compute the sum of pairwise distances for the given elements according to
+    the given dissimilarity matrix.
+
+    Parameters
+    ----------
+    elements: list or set
+        list or set of element IDs for which the sum of distances is computed
+    dm: skbio.stats.distance.DistanceMatrix
+        pairwise dissimilarity matrix.
+
+    Returns
+    -------
+    float:
+        the sum of all pairwise distances of dm for IDs in elements
+    '''
+
+    # some assertions
+    if type(dm) != DistanceMatrix:
+        raise TypeError('dm is not of type "DistanceMatrix"')
+    assert(len(elements) > 1)
+    assert(len(elements) <= len(dm.ids))
+    if len(set(elements)) != len(elements):
+        raise Exception('List of elements contain duplicates!')
+
+    # actual computation
+    return sum([dm[pair] for pair in set(itertools.combinations(elements, 2))])
```

Comment:
If it is really a distance matrix, this can significantly sped up by summing the upper or lower triangular matrix via [`numpy.tril`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.tril.html)

```python
In [1]: from skbio.stats.distance import randdm
In [2]: import numpy as np
In [3]: dm = randdm(100)
In [4]: elements = range(100)
In [5]: %timeit sum([dm[pair] for pair in set(itertools.combinations(elements, 2))])
100 loops, best of 3: 8.89 ms per loop

In [6]: %timeit np.tril(dm.data).sum()
The slowest run took 6.49 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 36.3 µs per loop

```